### ISSUE TITLE
request-based remote completion for RMA

### ISSUE THREAD

**Issue by mpiforumbot** [mpi-forum/mpi-forum-historic#398]
**Originally by _jhammond_ on 2013-10-31 15:23:43 -0500**

---

Request-based remote completion is useful for models that want fine-grain synchronization throughout.

See attachments for details.

**Comment by mpiforumbot** [comment]
**Originally by _rsthakur_ on 2013-12-09 19:20:07 -0600**

---

What is the critical use case to add these functions

MPI_RRPUT\
MPI_RRACCUMULATE\
MPI_RRRGET_ACCUMULATE  (with 3 request handles)

(There is a typo on pg 438: MPI_RRR_GETACCUMULATE)

**Comment by mpiforumbot** [comment]
**Originally by _jhammond_ on 2013-12-10 11:41:39 -0600**

---

**Attachment added:** [`one-side-2.ticket398.diff`](https://github.com/mpiforumbot/mpi-forum-historic/raw/master/attachments/398/one-side-2.ticket398.diff) (10.3 KiB)
patch for LaTeX source

**Comment by mpiforumbot** [comment]
**Originally by _jhammond_ on 2013-12-10 11:43:25 -0600**

---

**Attachment added:** [`mpi31-report.pdf`](https://github.com/mpiforumbot/mpi-forum-historic/raw/master/attachments/398/mpi31-report.pdf) (2707.6 KiB)
PDF of the MPI standard with modifications

**Comment by mpiforumbot** [comment]
**Originally by _jhammond_ on 2013-12-10 15:56:39 -0600**

---

Summary of working group (WG) discussion:
- The justification of fine-grain remote completion was determined to be weak by multiple members of the WG.  The two uses were stated to be CAF 2.0 and Charm++.  Members of the WG could not come up with a strong performance motivation i.e. could not come up with a hardware scenario where separation of remote completion would measurably improve performance.
- Brian did not like multiple requests and proposed instead to have a single request for a new function that would have end-to-end completion semantics, i.e. the completion of the request would correspond to **both** local and remote completion - no separation of local and remote completion is available.
- Pavan and others believe that nonblocking flush is sufficient to address the major issue at hand, which is that blocking flush causes problems.
- Jim and others believe that returning a request from `IWIN_FLUSH` and then testing on it is the wrong way to do this and proposes instead to add `WIN_TEST`, which has different semantics than `IWIN_FLUSH`.  In particular, there is no request to be managed.  However, since the `WIN_TEST` function already exists, we might need a new name for the proposed feature here so as to avoid confusion with PSCW (i.e. PSCT) usage.
- Jeff, Brian and possibly Jim thought that there might be some value in adding a SHMEM-like `PUT` to the MPI standard, which is a Put that blocks on local completion so as to eliminate the need for two function calls to achieve this effect.  This feature was determined to be out-of-scope w.r.t. this ticket and in need of its own ticket if the WG wishes to pursue it.

Thus, the WG concludes that we need to create a ticket for nonblocking flush, but only once we resolve the debate between a request-returning `IWIN_FLUSH` and Jim's `WIN_TEST`, which would have semantics that might not be exactly like any function already in the MPI standard.

**Comment by mpiforumbot** [comment]
**Originally by _rsthakur_ on 2015-06-03 13:21:03 -0500**

---

From the June 2015 Forum meeting: Jeff will retire this ticket.

**Comment by mpiforumbot** [comment]
**Originally by _jhammond_ on 2015-06-03 13:23:02 -0500**

---

WG had no objections to the desire by the reporter to retire this ticket.  The rationale is that there is likely a better way to do this (nonblocking flush) that is far more likely to pass.
